### PR TITLE
wallet: drop change to fee if amount is less than dust

### PR DESCRIPTION
**Pull request by Ayush170-Future**
This commit refactors the `CoinSelector's` `select()` method to properly handle the calculation of change cost. The change amount is now checked against the cost of change, and if the change is less than the cost, the cost is added to the transaction fee.

Solves: Issue [#26](https://github.com/Bitshala-Incubator/silent-pay/issues/26)

### CODE REVIEW

**Review by emjshrx — COMMENTED** [review]
Can we add a test

**Comment by Ayush170-Future** [comment]
> Can we add a test

Yes. I'll fix the lint and add a test in the next force push.

**Comment by Ayush170-Future** [comment]
Force pushed:

- Hopefully fixes Lint
- Adds a unit test for CoinSelector

File: src/wallet/coin-selector.ts [hunk 1]
```diff
@@ -36,20 +36,39 @@ export class CoinSelector {
...
+            change = 0;
+        }
 
         return {
             coins: selected.map((index) => pointers[index].coin),
             change,
         };
     }
 
+    costOfChange() {
```
Comment (theanmolsharma) [inline src/wallet/coin-selector.ts:59]: let's take advantage of TS syntactic sugar
```suggestion
    get costOfChange() {
```

FYI after you do this, you'd need to update the usages as well

File: src/wallet/coin-selector.ts [hunk 2]
```diff
@@ -36,20 +36,39 @@ export class CoinSelector {
...
+        // Calculate the cost of change
+        const LONG_TERM_FEERATE = 5; // in sats/vB
+        const outputSize = 31; // P2WPKH output is 31 B
+        const inputSizeOfChangeUTXO = 68.0; // P2WPKH input is 68.0 vbytes
```
Comment (theanmolsharma) [inline src/wallet/coin-selector.ts:63]: Add comments explaining the breakdown at byte level

File: src/wallet/coin-selector.ts [hunk 3]
```diff
@@ -36,20 +36,39 @@ export class CoinSelector {
...
 
         return {
             coins: selected.map((index) => pointers[index].coin),
             change,
         };
     }
 
+    costOfChange() {
+        // Calculate the cost of change
+        const LONG_TERM_FEERATE = 5; // in sats/vB
```
Comment (theanmolsharma) [inline src/wallet/coin-selector.ts:61]: this should be a static constant defined in the class

**Review by theanmolsharma — CHANGES_REQUESTED** [review]
will continue reviewing after you make these changes

**Comment by Ayush170-Future** [comment]
Force pushed [3c1983e](https://github.com/Bitshala-Incubator/silent-pay/commit/3c1983e944a543ac85aaf804df06bbf88add490b) addressing requested changes.

File: test/coinselector.spec.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        coins = [];
+        for (let i = 0; i < 10; i++) {
+            const value = Math.random() * 1000;
+            const coin = new Coin({ value });
+            coins.push(coin);
+            totalBalance += coin.value - coin.estimateSpendingFee(feeRate);
+        }
+    });
+
+    it('should generate cost of change', () => {
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:25]: For this particular test, I think you should calculate the cost of change by hand and then match it with the returned values.

Comment (theanmolsharma) [reply test/coinselector.spec.ts:25]: In summary, just match the output of `coinSelector.costOfChange` with the calculated value.

File: test/coinselector.spec.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        // Calculate fees
+        let fees = transaction.virtualSize() * feeRate;
+        fees += 31 * feeRate; // ChangeOutputFee: P2WPKH is 31 B
+        const finalTxOutValue = txOutValue - fees;
+
+        // Set the transaction output value
+        transaction.outs[0].value = finalTxOutValue;
+    });
+
+    it('should add change to fee if it is dust', () => {
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:49]: The calculations needed for this test should be done in this test itself. I see that you have a `beforeAll()` and some logic in previous tests, which according to me is misplaced. So I think you should move everything you have in `beforeAll()` and the tx creation part from previous test to this `it()`.

File: test/coinselector.spec.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        expect(typeof costOfChange).toBe('number');
+
+        // Calculate transaction output value
+        const txOutValue: number = Math.floor(totalBalance - costOfChange);
+
+        // Create a new transaction
+        transaction = new Transaction();
+        // Add an output with random scriptPubKey: We are only concered about its value
+        transaction.addOutput(
+            Buffer.from(validTransactions[0].raw.outs[0].script, 'hex'),
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:36]: instead of this, generate address randomly

File: test/coinselector.spec.ts
Comment (theanmolsharma) [inline test/coinselector.spec.ts:1]: Some thoughts on how `it('should add change to fee if it is dust',` should look like:
- create coins of random value that you pass to for selection.
- create and pass a tx with one output whose `value = totalBalance - 100`

All this logic should be part of the test body.

**Review by theanmolsharma — CHANGES_REQUESTED** [review]

**Comment by Ayush170-Future** [comment]
Force pushed [a9cf64d](https://github.com/Bitshala-Incubator/silent-pay/commit/a9cf64db7272b7a32102314b8fecfa844e8cb287) refactoring the unit test and creating random `script` for the test transaction output.

File: src/wallet/coin-selector.ts [hunk 7]
```diff
@@ -36,20 +37,57 @@ export class CoinSelector {
...
+        // - Total:
+        //   32 + 4 + 1 + 4 + (1 + 1 + 72 + 1 + 33) / 4 = 68 vbytes
+        const inputSizeOfChangeUTXO = 68.0;
+
+        const costOfChangeOutput = outputSize * this.feeRate;
+        const costOfSpendingChange =
+            inputSizeOfChangeUTXO * CoinSelector.LONG_TERM_FEERATE;
+        const costOfChange = costOfChangeOutput + costOfSpendingChange;
+
+        return costOfChange;
```
Comment (emjshrx) [inline src/wallet/coin-selector.ts:88]:
```suggestion
        return costOfChangeOutput + costOfSpendingChange;
```

File: src/wallet/coin-selector.ts [hunk 8]
```diff
@@ -36,20 +37,57 @@ export class CoinSelector {
...
+        //   - sequence (4 bytes)
+        //   - WITNESS STACK:
+        //     - item count (1 byte)
+        //     - signature length (1 byte)
+        //     - signature (71 or 72 bytes)
+        //     - pubkey length (1 byte)
+        //     - pubkey (33 bytes)
+        // - Total:
+        //   32 + 4 + 1 + 4 + (1 + 1 + 72 + 1 + 33) / 4 = 68 vbytes
+        const inputSizeOfChangeUTXO = 68.0;
```
Comment (emjshrx) [inline src/wallet/coin-selector.ts:81]: nit: `68`? Since byte sizes can never be decimal?

File: test/coinselector.spec.ts
Comment (emjshrx) [reply test/coinselector.spec.ts:1]: Tests should not have random values imo, as they lead to flaky tests. we should craft the test cases such that every scenario is covered.

File: test/coinselector.spec.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
+import { Coin } from '../src/wallet/coin';
+import { CoinSelector } from '../src/wallet/coin-selector.ts';
+import { Transaction } from 'bitcoinjs-lib';
+
+describe('CoinSelector', () => {
+    let coinSelector: CoinSelector;
+    const feeRate: number = 5;
```
Comment (emjshrx) [inline test/coinselector.spec.ts:7]: nit: typescript will infer the type so `const feeRate = 5;`

File: test/coinselector.spec.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    it('should generate correct cost of change', () => {
+        const outputSize = 31;
+        const inputSizeOfChangeUTXO = 68;
+
+        // Calculate cost of change based on output size and input size
+        const costOfChangeOutput = outputSize * feeRate;
+        const costOfSpendingChange =
+            inputSizeOfChangeUTXO * CoinSelector.LONG_TERM_FEERATE;
+        const correctCostOfChange = costOfChangeOutput + costOfSpendingChange;
```
Comment (emjshrx) [inline test/coinselector.spec.ts:21]: We dont need to do the calculation in the test. we can just have `expectedCostOfChange = value`

File: test/coinselector.spec.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+        // Get the calculated cost of change from the CoinSelector
+        const costOfChange = coinSelector.costOfChange;
+
+        // Assert the correctness of the CoinSelector result
+        expect(correctCostOfChange).toBe(costOfChange);
+    });
```
Comment (emjshrx) [inline test/coinselector.spec.ts:28]: nit 
```suggestion
        // Get the calculated cost of change from the CoinSelector
        const receivedCostOfChange = coinSelector.costOfChange;

        // Assert the correctness of the CoinSelector result
        expect(receivedCostOfChange).toBe(expectedCostOfChange);
    });
```

File: test/coinselector.spec.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        let totalBalance: number = 0;
+
+        // Generate random coins and calculate total balance
+        for (let i = 0; i < 20; i++) {
+            const value = Math.random() * 1000;
+            const coin = new Coin({ value });
+            coins.push(coin);
+            // Calculate total balance by subtracting spending fee
+            totalBalance += coin.value - coin.estimateSpendingFee(feeRate);
+        }
```
Comment (emjshrx) [inline test/coinselector.spec.ts:41]: this should be a constant list of coins imo. and we iterate through each set by something like it.each(). https://elfi-y.medium.com/reduce-boilerplate-test-code-with-jest-it-each-30a0eec9776d

File: test/coinselector.spec.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+        // Create a new transaction
+        const transaction: Transaction = new Transaction();
+        // Add an output with random scriptPubKey, we only care about its value
+        const randomString = [...Array(50)]
+            .map(() => Math.random().toString(36)[2])
+            .join('');
+        transaction.addOutput(Buffer.from(randomString, 'hex'), 1);
```
Comment (emjshrx) [inline test/coinselector.spec.ts:49]: Can we not have randomness and constant list of strings?

File: test/coinselector.spec.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+        // Calculate transaction fees
+        let fees = transaction.virtualSize() * feeRate;
+        fees += 31 * feeRate; // Additional fee for change output: P2WPKH is 31 bytes
+        const txOutValue =
+            Math.floor(totalBalance - coinSelector.costOfChange) - fees;
+
+        // Set the transaction output value
+        transaction.outs[0].value = txOutValue;
```
Comment (emjshrx) [inline test/coinselector.spec.ts:58]: If we use fixed test cases we dont need to do these calculations. We can have two test cases, one where change is dropped and one where it isnt. Isnt that enough @theanmolsharma

File: test/coinselector.spec.ts (hunk 13 above)
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:49]: Yes, I agree. Though this is just a place holder, the output script has no impact on the coin-selector file.

File: test/coinselector.spec.ts (hunk 11 above)
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:28]: Will fix this in the next force push.

File: test/coinselector.spec.ts (hunk 9 above)
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:7]: Okay.

File: test/coinselector.spec.ts (hunk 14 above)
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:58]: I can implement that as soon as Anmol reviews. Meanwhile, I'll force-push addressing your other concerns.

**Comment by Ayush170-Future** [comment]
Force pushed [0e8a816](https://github.com/Bitshala-Incubator/silent-pay/commit/0e8a816875ee051e8aff2da01e02dcfffe87d588), which changed the random data into constant values.

- I agree with @emjshrx that using random values in unit testing could lead to non-deterministic behavior. Therefore, I ran my code and recorded all the values it produced. I then hard-coded this static data to ensure deterministic results every time.

- I also covered both cases: when the `change > dust` and when the `change <= dust`.

- Fixed the `nit`s

**Review by theanmolsharma — CHANGES_REQUESTED** [review]
I think this should be the final set of changes required before merging. Nice work and thanks for your contribution.

File: test/coinselector.spec.ts [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+        );
+
+        // Assert that the selected coins' size is greater than 0
+        expect(selectedCoins.length).toBeGreaterThan(0);
+
+        // The above calculations ensure that the change is always less than the dust
+        expect(change).toBe(0);
+    });
+
+    it('change should be considered when greater than dust', () => {
```
[the remark this answers is not in the document]
Comment (emjshrx) [reply test/coinselector.spec.ts:66]: Both these tests are the same, just different values. Lets use `it.each()`. https://elfi-y.medium.com/reduce-boilerplate-test-code-with-jest-it-each-30a0eec9776d

[the remark this answers is not in the document]
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:66]: Nice. Will use `it.each()`

**Comment by Ayush170-Future** [comment]
Force pushed [ed44b85](https://github.com/Bitshala-Incubator/silent-pay/commit/ed44b85ee342c3f14c0d40f69b748e4583e75ecb)

- Used `it.each()`
- Added comments for justifying my calculations
- Replaced less than dust values in the coins with bigger values (more than `1000`)
- Rebased on the main

**Review by emjshrx — APPROVED** [review]
ack ed44b85ee342c3f14c0d40f69b748e4583e75ecb. Lets also wait for @theanmolsharma's approval

File: test/coinselector.spec.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+            expectedChange: 0, // Change will be zero when less than dust.
+            expectedCoinCount: 10, // All the coins are going to be selected.
+        },
+        {
+            testName: 'change should be considered when greater than dust',
+            testCoinValues: [5000, 4000],
+            txOutValue: 1000, // Less output so that the change is greater than dust.
+            expectedChange: 2185, // expectedChange = totalSelectedCoin - (txOutValue + transactionFees).
+            expectedCoinCount: 1, // One Coin will suffice for a 1000 sats output for this case.
+        },
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:36]: Two things:
1. In the second test case, if UTXO with value 4000, gets selected then the test will fail. It is not getting selected rn because the coin selection is deterministic but once we implement other selection algorithms, this might change. So only use 1 UTXO here or see the amount higher so that both get selected. An interesting case would be when you set `txOutValue` to be 5000.
2. I don't think we need `expectedCoinCount`, we should just be able to use `testCoinValues.length`.

File: test/coinselector.spec.ts [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                // Calculate the total balance by subtracting the spending fee.
+                totalBalance += coin.value - coin.estimateSpendingFee(feeRate);
+            }
+            
+            // Create a new Transaction
+            const transaction: Transaction = new Transaction();
+
+            // Fixed fees for the transaction calculated based on the total size of the transaction and the change output size.
+            // transactionFees = (virtualSize + changeOutputSize) * feeRate, P2WPKH output is 31 bytes.
+            const transactionFees = 250;
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:56]: how is the fee for both transactions the same? Both transactions have different number of inputs, so fee for both of them should be different.
Something is wrong here.

File: test/coinselector.spec.ts [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+            // txOutValue is not defined for the first test in order to demonstrate its calculation.
+            if (!txOutValue) {
+                // Fixed transaction output value that ensures the cost of change is dust.
+                txOutValue =
+                    totalBalance -
+                    transactionFees -
+                    coinSelector.costOfChange +
+                    1;
+            }
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:66]: I don't think this is the right approach, either define it in the test vector and add a comment for explanation or do this for both cases.

File: test/coinselector.spec.ts [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+            // Set the transaction output value with a random output script.
+            // Random hex string of 20 bytes and `0014${genRanHex}` will be a valid P2WPKH script.
+            const genRanHex = (size: number) =>
+                [...Array(size)]
+                    .map(() => Math.floor(Math.random() * 16).toString(16))
+                    .join('');
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:73]: Let's not do this and use `crypto.randomBytes()` instead

File: test/coinselector.spec.ts [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+            const { coins: selectedCoins, change } = coinSelector.select(
+                coins,
+                transaction,
+            );
+
+            // Assert that the selected coins' size meets the expected count.
+            expect(selectedCoins.length).toBe(expectedCoinCount);
+
+            // Assert the change against the expected change condition.
+            expect(change).toBe(expectedChange);
```
Comment (theanmolsharma) [inline test/coinselector.spec.ts:89]: nit: remove empty line
```suggestion
            // Assert that the selected coins' size meets the expected count.
            expect(selectedCoins.length).toBe(expectedCoinCount);
            // Assert the change against the expected change condition.
            expect(change).toBe(expectedChange);
```

**Review by theanmolsharma — CHANGES_REQUESTED** [review]
I think we can merge after the comments are addressed.

File: test/coinselector.spec.ts (hunk 17 above)
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:56]: I don't think so, `transactionFees` account for these calculations in the coin-selector. Here, the inputs have not been added to the transaction yet. If you do a `console.log()` for the fees calculations in lines `26 to 28`, it gives the same value for both tests. My assumption was that it only considers the number of outputs, and there is only one in both tests.

https://github.com/Bitshala-Incubator/silent-pay/blob/main/packages/wallet/src/coin-selector.ts#L26

File: test/coinselector.spec.ts (hunk 16 above)
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:36]: Thanks for this edge case. I'll edit the file with only one UTXO.

File: test/coinselector.spec.ts (hunk 17 above)
Comment (Ayush170-Future) [reply test/coinselector.spec.ts:56]: Anyway, I've removed the `transactionFees` and set static values for `txOutValue` in both tests. Therefore, we don't have to deal with transaction fees. Additionally, since all coins are selected in both tests, this approach won't pose any problems in the future, even when more coin selection algorithms are implemented.

**Comment by Ayush170-Future** [comment]
Force pushed [4589fff](https://github.com/Bitshala-Incubator/silent-pay/commit/4589ffff7b72453b4676e4dc3119cd34c473600c) addressing reviews and rebasing to the main.

File: packages/wallet/test/coinselector.spec.ts [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+            });
+
+            // Create a new Transaction
+            const transaction: Transaction = new Transaction();
+
+            // Set the transaction output value with a random output script.
+            // Random 20 bytes and `0014${randomBytes.toString()}` will be a valid P2WPKH script.
+            const randomBytes = crypto.randomBytes(20);
+            transaction.addOutput(
+                Buffer.from(`0014${randomBytes.toString('hex')}`), // Ensure the first parameter passed to addOutput is a Buffer.
```
Comment (theanmolsharma) [inline packages/wallet/test/coinselector.spec.ts:58]: nit
```suggestion
            // Random 20 bytes and `0014${crypto.randomBytes(20).toString()}` will be a valid P2WPKH script.
            transaction.addOutput(
                Buffer.from(`0014${crypto.randomBytes(20).toString('hex')}`), // Ensure the first parameter passed to addOutput is a Buffer.
```

**Review by theanmolsharma — APPROVED** [review]
One last nit and then I'll merge. (Ik I've been saying this for the past 3 reviews, but I promise to merge after this)
Awesome work @Ayush170-Future

**Comment by Ayush170-Future** [comment]
Done!!

Addressed the `nit`

**Review by theanmolsharma — APPROVED** [review]
ACK 991c16d2bd3841294dbf9a02495c7d6e30443692

**Comment by theanmolsharma** [comment]
@Ayush170-Future you need to sign the commits so that we can merge. I've enabled branch protection rules that only allow signed commits to be merged in main. It's a good FOSS practice to sign commits.
If you don't have a GPG setup now, I can sign the commits for this PR, but all commits in the future must be signed.

**Comment by Ayush170-Future** [comment]
> @Ayush170-Future you need to sign the commits so that we can merge. I've enabled branch protection rules that only allow signed commits to be merged in main. It's a good FOSS practice to sign commits. If you don't have a GPG setup now, I can sign the commits for this PR, but all commits in the future must be signed.

Yes, please sign the commits in this PR, and thanks for letting me know, I'll set it up and sign my commits from now on.
